### PR TITLE
Feature extra pages

### PR DESCRIPTION
**Pull request by AyAggarwal**
NOTE: this is blocked #224, but creating this anyways

### CODE REVIEW

**Review by jasonpaulos — APPROVED** [review]
Pending tests passing, looks great!

File: src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java [hunk 1]
```diff
@@ -71,4 +73,16 @@ public T globalStateSchema(StateSchema globalStateSchema) {
         this.globalStateSchema = globalStateSchema;
         return (T) this;
     }
+
+    /**
+     * extraPages allows you to rent extra pages of memory for the application. Each page is 2048 bytes of shared
+     * memory between approval and clear state programs. extraPages parameter must be an integer between 0 and 3 inclusive.
+     */
+    public T extraPages(Long extraPages) {
+        if (extraPages < 0 || extraPages > 3) {
```
Comment (jdtzmn) [inline src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java:82]:
```suggestion
        if (extraPages == null || extraPages < 0 || extraPages > 3) {
```
extraPages can be null and we'd still want the same exception to appear

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java [hunk 2]
```diff
@@ -150,6 +150,10 @@
     @JsonProperty("apsu")
     public TEALProgram clearStateProgram = null;
 
+    @JsonProperty("apep")
+    public Long extraPages = 0L;
+
+
```
Comment (jdtzmn) [inline src/main/java/com/algorand/algosdk/transaction/Transaction.java:156]:
```suggestion
```
nit: extra newline

**Review by jdtzmn — CHANGES_REQUESTED** [review]
Two small changes and then I'll approve.

File: src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java (hunk 1 above)
Comment (jasonpaulos) [reply src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java:82]: @jdtzmn good catch! Alternative suggestion: change the argument from `Long` to `long`.

Comment (AyAggarwal) [reply src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java:82]: I'll go with jacob's to match the other variables which are Long

**Review by winder — CHANGES_REQUESTED** [review]
The eval delta tests are for testing Indexer, and the Java SDK is the only one that implements them. Is something else going to cover these tests?

Is it possible these tests fail because Indexer changes are required?

https://github.com/algorand/algorand-sdk-testing/blob/master/features/integration/evaldelta.feature#L3

File: src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java [hunk 3]
```diff
@@ -169,11 +169,14 @@ public T assetName(String assetName) {
     }
 
     /**
-     * Set url. This value must be between 0 and 32 characters (inclusive).
+     * Set url. This value must be between 0 and 96 characters (inclusive).
      * @param url The asset url.
      * @return this builder.
      */
     public T url(String url) {
+        if (url.length() > 32) {
```
Comment (winder) [inline src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java:177]:
```suggestion
        if (url != null && url.length() > 96) {
```

File: src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java [hunk 4]
```diff
@@ -169,11 +169,14 @@ public T assetName(String assetName) {
...
+            throw new IllegalArgumentException("url must be between 0 and 32 characters (inclusive).");
```
Comment (winder) [inline src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java:178]:
```suggestion
            throw new IllegalArgumentException("url must be between 0 and 96 characters (inclusive).");
```

File: Makefile [hunk 5]
```diff
@@ -2,7 +2,7 @@ unit:
 	mvn test -Dcucumber.filter.tags="@unit.offline or @unit.algod or @unit.indexer or @unit.rekey or @unit.indexer.rekey or @unit.transactions or @unit.responses or @unit.applications or @unit.dryrun or @unit.tealsign or @unit.responses.messagepack or @unit.responses.231 or @unit.responses.messagepack.231"
 
 integration:
-	mvn test -Dcucumber.filter.tags="@algod or @assets or @auction or @kmd or @send or @template or @indexer or @rekey or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @applications.evaldelta or @indexer.231"
+	mvn test -Dcucumber.filter.tags="@algod or @assets or @auction or @kmd or @send or @template or @indexer or @rekey or @applications.verified or @applications or @compile or @dryrun or @indexer.applications or @indexer.231"
```
Comment (winder) [inline Makefile:5]: Why remove this test?

Comment (AyAggarwal) [reply Makefile:5]: I added it back and made a new fix, initializing extraPages = 0L in the apptxncreatebuilder

File: src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java (hunk 4 above)
Comment (AyAggarwal) [reply src/main/java/com/algorand/algosdk/builder/transaction/AssetCreateTransactionBuilder.java:178]: I have another PR for this, is it cool if I keep these 32 on this PR?

File: Makefile (hunk 5 above)
Comment (shiqizng) [reply Makefile:5]: still missing @applications.evaldelta

Comment (AyAggarwal) [reply Makefile]: added

File: src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java [hunk 6]
```diff
@@ -7,6 +7,7 @@
 public class ApplicationCreateTransactionBuilder<T extends  ApplicationCreateTransactionBuilder<T>> extends ApplicationUpdateTransactionBuilder<T> {
     private StateSchema localStateSchema;
     private StateSchema globalStateSchema;
+    private Long extraPages = 0L;
```
Comment (AyAggarwal) [inline src/main/java/com/algorand/algosdk/builder/transaction/ApplicationCreateTransactionBuilder.java:10]: @shiqizng , this is what I added to initialize default to 0

File: src/main/resources/langspec.json [hunk 7]
```diff
@@ -1 +1 @@
...
... params field index}","Groups":["State Access"]},{"Opcode":120,"Name":"min_balance","Args":"U","Returns":"U","Cost":1,"Size":1,"Doc":"get minimum required balance for the requested account specified by Txn.Accounts[A] in microalgos. A is specified as an account index in the Accounts field of the ApplicationCall transaction, zero index means the sender. Required balance is affected by [ASA](https://developer.algorand.org/docs/features/asa/#assets-overview) and [App](https://developer.algorand.org/docs/features/asc1/stateful/#minimum-balance-requirement-for-a-smart-contract) usage. When creating or opting into an app, the minimum balance grows before the app code runs, therefore the increase is visible there. When deleting or closing out, the minimum balance decreases after the app executes.","Groups":["State Access"]},{"Opcode":128,"Name":"pushbytes","Returns":"B","Cost":1,"Size":0,"Doc":"push the following program bytes to the stack","ImmediateNote":"{varuint length} {bytes}","Groups":["Loading Values"]},{"Opcode":129,"Name":"pushint","Returns":"U","Cost":1,"Size":0,"Doc":"push immediate UINT to the stack as an integer","ImmediateNote":"{varuint int}","Groups":["Loading Values"]}]}
```
Comment (shiqizng) [inline src/main/resources/langspec.json:1]: make sure langspec file is the same as what's in develop

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java [hunk 8]
```diff
@@ -42,9 +42,9 @@ public static TEALProgram loadTEALProgramFromFile(String file) {
...
+    @When("I build an application transaction with operation {string}, application-id {long}, sender {string}, approval-program {string}, clear-program {string}, global-bytes {long}, global-ints {long}, local-bytes {long}, local-ints {long}, app-args {string}, foreign-apps {string}, foreign-assets {string} app-accounts {string}, fee {long}, first-valid {long}, last-valid {long}, genesis-hash {string}, extra-pages {long}")
+    @When("I build an application transaction with operation {string}, application-id {long}, sender {string}, approval-program {string}, clear-program {string}, global-bytes {long}, global-ints {long}, local-bytes {long}, local-ints {long}, app-args {string}, foreign-apps {string}, foreign-assets {string}, app-accounts {string}, fee {long}, first-valid {long}, last-valid {long}, genesis-hash {string}, extra-pages {long}")
```
Comment (shiqizng) [inline src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:46]: why both @When(...) are needed?

File: src/test/java/com/algorand/algosdk/logic/TestLogic.java [hunk 9]
```diff
@@ -264,4 +264,4 @@ public void testCheckProgramTealV3() throws Exception {
             assertThat(valid).isTrue();
         }
     }
-}
+}
```
Comment (shiqizng) [inline src/test/java/com/algorand/algosdk/logic/TestLogic.java:267]: nit: restore the new line

File: src/main/resources/langspec.json (hunk 7 above)
Comment (AyAggarwal) [reply src/main/resources/langspec.json:1]: yep, copied it directly

File: src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java (hunk 8 above)
Comment (AyAggarwal) [reply src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:45]: yep, copied it directly

Comment (winder) [reply src/test/java/com/algorand/algosdk/cucumber/shared/TransactionSteps.java:45]: This isn't right, if you copied it directly then one of the feature files is wrong and the missing comma should be added

**Pull request merged**